### ISSUE TITLE
GaussianBeam2DSource Bug Cases Found

### ISSUE THREAD

**Issue by hammy4815** [NanoComp/meep#2356]
A few bugs were discovered regarding the new GaussianBeam2DSource implementation.

1. Using materials other than air produces slight issues after creating the equivalent source. 

For example, as can be seen below, after changing the refractive index to 1.5, small issues arise and back propagation below the source region now shows up. I am unsure at this moment if this is an issue in the creation of the beam fields themselves, or in the call to equivalent source, but it appears that a factor of $n$ or $\epsilon$ is either missing or accidentally added. 

<img width="460" alt="Screenshot 2022-12-27 at 2 42 48 PM" src="https://user-images.githubusercontent.com/51729722/209731419-50e148ba-6b2d-4c84-918d-e39957ff9927.png">

Currently, as given [here](https://arxiv.org/pdf/1301.5366.pdf) in equation 4.4, the equivalent current sources do not contain a factor of $\epsilon$ or $\mu$, matching how it is currently implemented in MEEP. 

<img width="702" alt="Screenshot 2022-12-27 at 2 51 39 PM" src="https://user-images.githubusercontent.com/51729722/209732827-c3105d39-ca10-4923-9247-3a65e1a79d2d.png">

2. Under stronger index changes, or a significant wavelength change, an overflow occurs in the normalization of the fields before calling equivalent source, and the time fields remain zero. 

<img width="442" alt="Screenshot 2022-12-27 at 2 46 39 PM" src="https://user-images.githubusercontent.com/51729722/209731641-0f2ea77c-15a0-45bc-8a20-d1520760cda6.png">

This is because calling scipy.special.hankel produces enormous results (on the order of 1e120). When the beam waist is similar in size to the effective wavelength, the order of magnitude is low enough to avoid overflows. We likely need a clever way to fix this issue at this step (calling the hankel function).

**Comment by oskooi** [comment]
The refractive index $n$ (of free space) is implicit in the fields $E_+$ and $H_+$ .There is no need for the index to be accounted for separately in the expression for the equivalent current.

Note that in the original implementation of the `GaussianBeam` object via the complex point source in #711 (and also the [analytic expression for the electric field of a Gaussian beam](https://en.wikipedia.org/wiki/Gaussian_beam#Mathematical_form)), the beam's wavevector $\vec{k}$ in free space includes the index $n$:

https://github.com/NanoComp/meep/blob/802257f72b84ad9d04985145cce5dd5d8db6b94c/src/sources.cpp#L604-L608

**Comment by hammy4815** [comment]
Just an update - I fixed issue #1. 

There was an extra factor of Z (impedance) on H during my normalization that shouldn't be there. 

https://github.com/NanoComp/meep/blob/d29eadcad29b6efa333fdafa6f59e3698872f1df/python/source.py#L902-L903

In addition, the resolution needs to be high enough to support higher epsilon. 

https://github.com/NanoComp/meep/blob/d29eadcad29b6efa333fdafa6f59e3698872f1df/python/source.py#L851-L856

Finally, I reformulated the way I use the polarization vector, because my previous method did not take into account the phases properly. This was the code before:

https://github.com/NanoComp/meep/blob/d29eadcad29b6efa333fdafa6f59e3698872f1df/python/source.py#L905-L914

Now, in green2d, the polarization vector used in the formulation matches the user provided polarization vector for the electric source:

```python
p = np.zeros(3, dtype=complex)
p[0] = beam_E0.x
p[1] = beam_E0.y
p[2] = beam_E0.z
pdotrhat = p[0] * rhat[0] + p[1] * rhat[1]
rhatcrossp = rhat[0] * p[1] - rhat[1] * p[0]
```

And for the magnetic source that is pointed orthogonal to the electric source:

```python
p = - np.array([
    - p[2] * np.imag(kdir),
    p[2] * np.real(kdir),
    p[0] * np.imag(kdir) - p[1] * np.real(kdir)
])
pdotrhat = p[0] * rhat[0] + p[1] * rhat[1]
rhatcrossp = rhat[0] * p[1] - rhat[1] * p[0]
```

I will create a PR for these changes, but am still trying to solve the final problem of the overflow. The hankel functions are returning nan's during overflow, which causes problems.

**Comment by hammy4815** [comment]
Hi guys, 

I have an update on the issue with nan's and infs for high index. Maybe I'm missing something here, but the problem seems more fundamental than a normalization issue.

Consider the beam with waist size $w_0 = 1 \mu m$, frequency $f = 1$, $\omega = 2\pi$, in a medium of $n = 6$ corresponding to a $k=2\pi 6 = 12\pi$:

The complex origin of the dipole is placed at $z0 = jkw_0^2/2=j6\pi$ which in two dimensions if propagating in the y direction is the vector $\mathbf{p_0}=[0,\pm  j6\pi]$ depending on the direction of $\mathbf{k}$. 

At the origin $\mathbf{r} = [0, 0]$, the radial distance $\mathbf{r}-\mathbf{p_0} = \pm j6\pi$. The argument of the hankel function is $kr= \pm12\pi * j6\pi =  \pmj72\pi^2 \approx  \pm710j$.

Passing these arguments into jv, yv, hankel1, and hankel2 shows that this argument is too (imaginarily) large and the bessel/hankel functions break down. This argument is located at the beam waist and should have maximal amplitude. 

```python
>>> yv(0, 710j)
(nan+nanj)
>>> yv(0, -710j)
(nan+nanj)
>>> jv(0, -710j)
(inf+nanj)
>>> jv(0, +710j)
(inf+nanj)
>>> hankel1(0, +710j)
0j
>>> hankel2(0, -710j)
(-0+0j)
>>> hankel1(0, -710j)
(nan+nanj)
>>> hankel1(0, +710j)
0j
```

Note that `hankel1 = jv + j * yv`
and `hankel2 = jv - j * yv`

If anyone has any insights or thoughts, I'd love to hear them.

**Comment by stevengj** [comment]
You can always use the exponentially scaled Hankel functions [`scipy.special.hankel1e`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.hankel1e.html) and [`scipy.special.hankel2e`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.hankel2e.html).  This lets you pull out the exponential factor analytically if you want to rescale.

**Comment by hammy4815** [comment]
> You can always use the exponentially scaled Hankel functions [`scipy.special.hankel1e`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.hankel1e.html) and [`scipy.special.hankel2e`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.hankel2e.html). This lets you pull out the exponential factor analytically if you want to rescale.

This works well. It appears to be working now that I switched to the exponential Hankel functions and am careful to scale and normalize correctly for the hankel1 vs hankel2. I will create a new PR with the changes and a description of how I am performing the proper scaling.

**Comment by stevengj** [comment]
Closed by #2363 and #2371
